src/Functions/formatDateTime.cpp

```diff
@@ -847,7 +861,8 @@ class FunctionFormatDateTimeImpl : public IFunction
         dst_data.resize(vec.size() * (result_size + 1));
         dst_offsets.resize(vec.size());
 
-        if constexpr (format_syntax == FormatDateTimeTraits::FormatSyntax::MySQL)
+        if (format_syntax == FormatDateTimeTraits::FormatSyntax::MySQL
```

Comment:
Hi Jake,

I thought about this a bit more ... in l. 861, the string output column is resized according to the computed (fixed) output width. String columns in ClickHouse are organized in two parts: 1. the "data part" which is an array of the 0-terminated raw strings (i.e. strings are separated by a zero byte). 2. the "offset part" which is an array of relative integer offsets into the data part.

In l. 904, the code writes for each row and each instruction into the previously determined position which is passed as `char * dest` into each instruction (e.g. `static size_t mysqlMonth(char * dest, Time source, UInt64, UInt32, const DateLUTImpl & timezone)`, l. 342). Instructions return how many characters were written into the output (and thus by how many characters `dest` is advanced before the instruction executes). For fixed-length formatters, this value will be the same across all rows. For variable length-formatters, it won't. This creates two problems:
1. Formatters `%M` and `%W` return "January" and "Monday" in the initial calculation of the width (l. 1053, l. 1053). Now imagine a simple example with format string `%M %W` and a single input row `toDate('2023-11-09')`. We will allocate 7 (January) + 6 (Monday) + 1 (trailing zero byte) bytes in the output string data part array. However, both instructions will actually write 8 (November) + 8 (Thursday) + 1 (trailing zero byte) into the output structure. That's a memory corruption. To solve this, the output width for `%M` and `%W` must be calculated based on the longest possible month/weekday (September, Wednesday).
2. Doing so creates another problem: If the actual input dates create output strings shorter than the previously calculated "worst-case" width (and that will be very often the case), then the result array will have a part at the end which was allocated but not written by any of the instructions. The data in this area will depend on whether the initial resize (l. 861) initializes the result array with zeros or not (in which case the data will be arbitrary garbage). To avoid that, we should shrink/resize the result array once more after all instructions ran to the end of the actually written data. EDIT: I am surprised to see that this happens already in l. 911 even though so far only fixed-size formatters were used.

Reply:
Hey Robert, (is it okay if I call you by that name?). Thanks for the detailed explanation 😊  That makes a lot of sense, my thoughts for the last few days have been to try and find a way to write the dynamic strings directly to the `out_template` but this led to a lot of code duplication and was just quite messy overall.

So I need to first ensure that I am using the largest possible string for  both the `M/W` formatters to guarantee that we have enough memory allocated *before* we call our instructions that write to the output (l.904).

Despite the resizing (at l. 911) we are still getting garbage out for any fixed formatters adjacent to a dynamic formatter. For example:
```cpp
select DATE_FORMAT(toDate32('2022-09-02'), '%r-%W' 
>>> 7;53Friday  // should be 12:00 AM-Friday
```
Does this mean that we need to find a way to resize as we write each instruction? Resizing the array after the *all* the results have been written will lead to data-loss somewhere along the entire string or do we look at moving up a layer and defining a new method altogether (alternative to `executeType`)? This would lead to quite a lot of duplicated code I imagine.

Lastly, this also means that the `W` change (already released) has introduced a bug into the codebase which wasn't caught by our tests, so I when/if I get this resolved I need to make sure I write new tests specifically for checking dynamic-formatters.

I must admit I am struggling a bit here but I'm learning a lot. Thanks as always for your patience 😊

Reply:
Hi Jake (yes, Robert is okay of course),

I don't have an explanation for the weird output, but I can perhaps debug tomorrow or so and check what goes wrong with `%r-%W`. In any case, a new top-level method next `executeType` would be quite ugly / add too much duplication, and I also don't think it is necessary. As far as I see, it is safe if each instruction (fixed or variable length) simply writes its output into the output structure which was allocated based on a worst-case length estimation. 

Good point on `W` ... yes, there should definitely be a test for that, thanks for noticing.

Reply:
Oh, and please don't feel pressured to finish this PR. Just let me/us know if we should continue, after all we are payed $ (in my case €) to work on ClickHouse.